### DISCUSSION TITLE
SQLalchemy warning of carthesian product when using jsonb_to_recordset -- correct or bogus warning?

### DISCUSSION THREAD

**Discussion by eveith** [sqlalchemy/sqlalchemy#8773] [Usage Questions]
[ Disclairmer: This is a cross-post from [StackOverflow](https://stackoverflow.com/questions/74346944/sqlalchemy-warning-of-carthesian-product-when-using-jsonb-to-recordset-correc) ]

Dear SQLalchemy community,

I need a bit of help with CTEs, a more complex query with `JSON` in PostgreSQL, and SQLalchemy's JOIN warning. I have an extended query constructed in SQLalchemy that queries the following table relations:

    experiment_run_phases --(1:n)-- agents --(1:n)-- muscle_actions

In the `muscle_actions` relation, there's an attribute `sensor_readings` that holds a list of JSON objects, so I'm using `jsonb_to_recordset` to extract the relevant fields/values to continue to work with it. The “steps” of the query are neatly formulated as separate CTEs.

Specifically, this first CTE retrieves all relevant rows from `muscle_actions`:

    q_mas = (
        sa.select(
            paldb.MuscleAction.id.label("ma_id"),
            paldb.MuscleAction.walltime.label("walltime"),
            paldb.ExperimentRunPhase.uid.label("phase_uid"),
            sa.literal_column(f"timestamp '{phase_start_dt}' + make_interval(secs => (muscle_actions.simtimes->'env'->>'simtime_ticks')::INTEGER)").label("simtime"),
            paldb.MuscleAction.sensor_readings.label("sensor_readings")
        )
        .join(
            paldb.MuscleAction.agent
        )
        .join(
            paldb.ExperimentRunPhase
        )
        .where(
            paldb.ExperimentRunPhase.id.in_(phase_ids),
            paldb.Agent.uid.in_(agent_names)
        )
        .order_by(
            paldb.MuscleAction.id
        )
    ).cte("q_muscle_actions")

The next now works on it to unpack the JSON objects:

    sensor_information_record = (
        sa.func.jsonb_to_recordset(
            sa.func.jsonb_path_query_array(
                sa.cast(q_mas.c.sensor_readings, sa.dialects.postgresql.JSONB),
                '$[*] ? (@.sensor_id like_regex "vm_pu")'
            )
        ).table_valued(
            sa.column("sensor_id", sa.Text),
            sa.column("sensor_value", sa.dialects.postgresql.JSONB),
        ).render_derived(name="sensor_information_record", with_types=True)
    )

Now that I have a record called `sensor_information_record` I simply want to use it: Each sensor reading gets a separate line (not extra columns). I.e., the first query retrieved records with the following attributes:

    ma_id(int),
    walltime(timestamp),
    phase_uid(varchar),
    simtime(timestamp),
    sensor_readings(JSON)

Using the second CTE, I now have the following record in addition:

    sensor_id(text),
    sensor_value(JSONB)

Finally, I combine the two:

    q = (
        sa.select(
            func.row_number().over(
                order_by=[
                    q_mas.c.ma_id,
                    q_mas.c.walltime,
                    sensor_information_record.c.sensor_id
                ]
            ).label("id"),
            q_mas.c.ma_id,
            q_mas.c.simtime,
            q_mas.c.phase_uid,
            sensor_information_record.c.sensor_id.label("sensor_id"),
            sa.cast(
                sensor_information_record.c.sensor_value["value"],
                sa.REAL
            ).label("sensor_value"),
            (
                sa.func.lag(
                    sa.cast(
                        sensor_information_record.c.sensor_value["value"],
                        sa.REAL
                    ),
                    1,
                    1.0
                ).over(
                    partition_by=sensor_information_record.c.sensor_id,
                    order_by=q_mas.c.simtime
                ) - sa.cast(
                    sensor_information_record.c.sensor_value["value"],
                    sa.REAL
                )
            ).label("delta_sensor_value")
        )
        .select_from(q_mas)
    ).cte("q_sensor_data")

… and now I can work further with `q`.

Now, SQLalchemy gives me numerous warnings with the same text:

> SELECT statement has a cartesian product between FROM element(s) "sensor_information_record" and FROM element "q_muscle_actions".  Apply join condition(s) between each element to resolve.

But even if I run it in the DB console with a subset of, say, initially 5 rows, and count it, I get the correct number of rows in the final result (e.g., with 5 sensor readings, I get 5*5=25 result records).  But I can understand that SQLalchemy's heuristic produces that warning because there's, in the end, a `FROM a, b` clause in the resulting query. Warnings nag, so I'd like to solve this.

1. Do I have a bug hidden in my query here that I do not understand?
2. How can I silence this warning?
3. How *could* I come up with a JOIN condition between the `q_muscle_actions` and `sensor_information_record` CTEs?

Thanks a lot in advance for any help!

Best,
Eric

**Comment by zzzeek** [comment]
> Now, SQLalchemy gives me numerous warnings with the same text:
> 
> > SELECT statement has a cartesian product between FROM element(s) "sensor_information_record" and FROM element "q_muscle_actions".  Apply join condition(s) between each element to resolve.
> 
> But even if I run it in the DB console with a subset of, say, initially 5 rows, and count it, I get the correct number of rows in the final result (e.g., with 5 sensor readings, I get 5*5=25 result records). But I can understand that SQLalchemy's heuristic produces that warning because there's, in the end, a `FROM a, b` clause in the resulting query. Warnings nag, so I'd like to solve this.
> 
>     1. Do I have a bug hidden in my query here that I do not understand?

I would say it's unlikely as you have tested it and ensured you are in fact lining things up

> 
>     2. How can I silence this warning?

other than using the warnings filter, you would need to either add an extra WHERE criteria that satisfies the path traveral, or construct a very succinct and clear [MCVE](http://stackoverflow.com/help/mcve) and submit a bug report here so that I can try to see if there's a way for the path traversal to pick up on what it is you're doing here (note that with a series of fragments, I generally dont really know the specifics of how this works as it's a lot of effort for me to piece together your above fragments into a complete and minimal picture, hence the MCVE)


> 
>     3. How _could_ I come up with a JOIN condition between the `q_muscle_actions` and `sensor_information_record` CTEs?

I would need a succinct, complete, single file MCVE so I can begin to understand how your query works.

**Comment by CaselIT** [comment]
> How could I come up with a JOIN condition between the q_muscle_actions and sensor_information_record CTEs?

the usual way of silencing it would be to join on True, so something like
`sa.select( ...).select_from(q_mas).join(sensor_information_record, sa.true())`

not sure if the db likes it, considering that `sensor_information_record` is a table valued function.
